### PR TITLE
Feature/as 2920 check sacd when issuing privilege tokens

### PR DESCRIPTION
**Pull request by LorranSutter**

### CODE REVIEW

File: internal/controllers/token_exchange.go [hunk 1]
```diff
@@ -92,6 +92,12 @@ func (t *TokenExchangeController) GetDeviceCommandPermissionWithScope(c *fiber.C
 		t.logger.Fatal().Err(err).Str("Contracts", pr.NFTContractAddress).Msg("Unable to initialize nft contract")
 		return fiber.NewError(fiber.StatusInternalServerError, "Could not connect to blockchain node")
 	}
+	s, err := t.ctmr.GetSacd(t.settings.ContractAddressSacd, client)
+	if err != nil {
+		t.logger.Fatal().Err(err).Str("Contracts", pr.NFTContractAddress).Msg("Unable to initialize nft contract")
```
Comment (elffjs) [inline internal/controllers/token_exchange.go:97]: This fatal is probably going to tear down the entire app, so no point in returning anything really. We should probably take out the fatal.

File: internal/controllers/token_exchange.go [hunk 2]
```diff
@@ -118,6 +124,15 @@ func (t *TokenExchangeController) GetDeviceCommandPermissionWithScope(c *fiber.C
 			return fiber.NewError(fiber.StatusForbidden, fmt.Sprintf("Address lacks privilege %d.", p))
 		}
 	}
+
+	res, err := s.HasPermissions(nil, common.HexToAddress(pr.NFTContractAddress), big.NewInt(pr.TokenID), *ethAddr, intArrayToBitArray(pr.Privileges, 256))
```
Comment (elffjs) [inline internal/controllers/token_exchange.go:128]: What feels bad about this is that I can't tell someone which permissions they don't have.

File: internal/controllers/token_exchange.go [hunk 3]
```diff
@@ -139,3 +154,16 @@ func (t *TokenExchangeController) GetDeviceCommandPermissionWithScope(c *fiber.C
 		Token: tk,
 	})
 }
+
+func intArrayToBitArray(indices []int64, length int) *big.Int {
+	bitArray := big.NewInt(0)
+
+	for _, index := range indices {
+		if index >= 1 && index <= int64(length) {
```
Comment (elffjs) [inline internal/controllers/token_exchange.go:162]: Is this good enough? I think I need `2*index` to be less than `length`.

File: internal/contracts/sacd/Sacd.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
+// Code generated - DO NOT EDIT.
+// This file is a generated binding and any manual changes will be lost.
+
+package contracts
```
Comment (elffjs) [inline internal/contracts/sacd/Sacd.go:4]: Package should be `sacd` if the folder's named that.

Comment (LorranSutter) [reply internal/contracts/sacd/Sacd.go:4]: Multiprivilege is capital case. Do we have a style guide for that?

File: internal/controllers/token_exchange.go (hunk 3 above)
Comment (LorranSutter) [reply internal/controllers/token_exchange.go:162]: The param I passed above was 256. But I removed that, so doesn't matter anymore

File: internal/controllers/token_exchange.go (hunk 1 above)
Comment (LorranSutter) [reply internal/controllers/token_exchange.go:97]: Removed from multiprivilege too

There is also `Fatal` in the client initialization above. Does it make sense to remove there too? Not sure what would be the best way to handle that in this case.

File: internal/controllers/token_exchange.go (hunk 2 above)
Comment (LorranSutter) [reply internal/controllers/token_exchange.go:128]: I changed to query one by one

Comment (LorranSutter) [reply internal/controllers/token_exchange.go:128]: I think none of that will be necessary once I include the sacd call in the NFT tbh

File: internal/contracts/sacd/Sacd.go (hunk 4 above)
Comment (elffjs) [reply internal/contracts/sacd/Sacd.go]: This is just writing any Go code for the last 14 years. There are lots of good style guides but I'm not sure they ever call this out.

File: internal/controllers/token_exchange.go (hunk 1 above)
Comment (elffjs) [reply internal/controllers/token_exchange.go]: That was a mistake too.

**Pull request merged**